src/models/kinflated_gamma_poisson_model.py

```diff
@@ -0,0 +1,633 @@
+            to be fit.
+          kmax:  Maximum number of values of the PMF that are allowed to be set
+            arbitrarily.
+          debug:  If True, prints debug information as models are fit.
+          extrapolation_multiplier:  Float.  If specified, then a penalty term is
+            introduced that penalizes models where the expected number of impressions
+            in the inventory is less than extrapolation_multiplier * the observed number
+            of impressions in the data.
+        """
+        if len(data) != 1:
+            raise ValueError("Exactly one ReachPoint must be specified")
+        if data[0].impressions[0] < 0.001:
+            raise ValueError("Attempt to create model with 0 impressions")
+        if data[0].impressions[0] < data[0].reach(1):
+            raise ValueError(
+                "Cannot have a model with fewer impressions than reached people"
+            )
+        self._reach_point = data[0]
+        self._kmax = kmax
+        self._debug = debug
+        self._fit_computed = False
+        self._extrapolation_multiplier = extrapolation_multiplier
+        self._extrapolation_value = extrapolation_multiplier * data[0].impressions[0]
+        if data[0].spends:
+            self._cpi = data[0].spends[0] / data[0].impressions[0]
+        else:
+            self._cpi = None
+
+    def _chi2_distance(
+        self,
+        h: npt.ArrayLike,
+        I: float,
+        N: float,
+        dist: KInflatedGammaPoissonDistribution,
+    ) -> float:
+        """Returns distance between actual and expected histograms.
+
+        Computes the metric
+
+            chi2(h, hbar) = \sum_i (h[i] - hbar[i])^2 / hbar[i]
+
+        where
+            hbar is the histogram that would be expected given parameters
+               I, N, dist
+
+        The first part of this objective formula is the chi-squared statistic
+        used for determining whether two sets of categorical samples are drawn
+        from the same underlying distribution.  Therefore, if the value returned
+        by this function is relatively low (say < 20), then there is reason to
+        believe that the Gamma-Poisson distribution adequately fits the data.
+
+        Args:
+          h: Actual frequency histogram.  h[i] is the number of people
+             reached exactly i+1 times, except for h[-1] which is the
+             number of people reached len(h) or more times.
+          I: Total number of impressions purchased.
+          N: Total number of unique users.
+          dist:   A KInflatedGammaPoisson distribution.
+        Returns:
+          chi^2 distance between the actual and expected histograms, plus
+          a term for weighting the difference in 1+ reach.
+        """
+        # Estimate total number of potential impressions
+        Imax = N * dist.expected_value()
+        if Imax <= I:
+            return np.sum(np.array(h) ** 2)
+
+        freqs = N * dist.kreach(np.arange(1, len(h)), I / Imax)
```

Comment:
We should have `I >= sum([(k + 1) * h[k] for k in range(len(h))`.  Correct me if I'm wrong.  
You may check if this condition is always satisfied.  If it is violated, will it introduce bug?

Reply:
If Imax <= I, then there will be an error in the computation of the histogram, because the probability that a given impression is selected is I / Imax, and this value would be greater than 1.  So, for that reason, it's important to check this boundary condition.  If I < sum([(k + 1) * h[k] for k in range(len(h)), then the result will be that the quality of the fit is poor and the value of the objective function is high.  It will not cause an error (to my knowledge).  So, in my view, this does not need to be checked separately.